modules/core/src/main/java/org/apache/ignite/internal/jdbc2/JdbcBinaryBuffer.java

```diff
@@ -0,0 +1,449 @@
+     */
+    public static JdbcBinaryBuffer createReadWrite(byte[] arr) {
+        return new JdbcBinaryBuffer(arr, 0, arr.length, false);
+    }
+
+    /**
+     * Create empty buffer which starts working in the read-write mode.
+     */
+    public static JdbcBinaryBuffer createReadWrite() {
+        return new JdbcBinaryBuffer(new byte[MIN_CAP], 0, 0, false);
+    }
+
+    /**
+     * Private constructor.
+     *
+     * @param arr The byte array to be wrapped.
+     * @param off The offset to the first byte to be wrapped.
+     * @param len The length in bytes of the data to be wrapped.
+     * @param isReadOnly The read-only flag.
+     */
+    private JdbcBinaryBuffer(byte[] arr, int off, int len, boolean isReadOnly) {
+        this.arr = arr;
+        this.off = off;
+        this.len = len;
+        this.isReadOnly = isReadOnly;
+    }
+
+    /**
+     * Create shallow read-only copy of this buffer.
+     */
+    public JdbcBinaryBuffer shallowCopy() {
+        return new JdbcBinaryBuffer(arr, off, len, true);
+    }
+
+    /**
+     * Provide InputStream through which the data can be read starting from the
+     * begining.
+     *
+     * <p>Stream is not limited meaning that it would return any new data
+     * written to the buffer after stream creation.
+     *
+     * @return InputStream instance.
+     */
+    public InputStream inputStream() {
+        return new BufferInputStream();
+    }
+
+    /**
+     * Get copy of the buffer data as byte array.
+     *
+     * @return Byte array containing buffer data.
+     */
+    public byte[] bytes() {
+        byte[] bytes = new byte[len];
+
+        read(0, bytes, 0, len);
+
+        return bytes;
+    }
+
+    /**
+     * Provide OutputStream through which the data can be written to buffer starting from
+     * the (zero-based) {@code pos} position.
+     *
+     * @param pos The zero-based offset to the first byte to be written. Must not be negative
+     *            or greater than total count of bytes in buffer.
+     *
+     * @return OutputStream instance.
+     */
+    OutputStream outputStream(int pos) {
+        return new BufferOutputStream(pos);
+    }
+
+    /**
+     * Provide InputStream through which the no more than {@code len} bytes can be read
+     * from buffer starting from the specified zero-based position {@code pos}.
+     *
+     * @param pos The zero-based offset to the first byte to be retrieved. Must not be negative
+     *            or greater than total count of bytes in buffer.
+     * @param len The length in bytes of the data to be retrieved. Must not be negative.
+     * @return InputStream instance.
+     */
+    InputStream inputStream(int pos, int len) {
+        return new BufferInputStream(pos, pos + len);
+    }
+
+    /**
+     * Truncate data in this buffer to specified length.
+     *
+     * @param len New length.
+     */
+    void truncate(int len) {
+        this.len = len;
+
+        reallocate(Math.max(MIN_CAP, len));
```

Comment:
Why we want to keep MIN_CAP size even if user ask to trunkcate data to `len`?

Reply:
> We never have buffer smaller then MIN_CAP.

How you come to that requirement?
I don't think we have it.

Reply:
In one of the previous comments you say about method which writes 1 byte to stream:

> We don't want to create byte[1] to write data.
> Let's create new array of reasonable size here.

So, MIN_CAP == 256 is looks reasonable.

***
But we can respect user here (but not in write method) and truncate to exact value.

And afterwards if user would start write increase at least for MIN_CAP